rc/parinfer.kak

```diff
@@ -1,37 +1,77 @@
-declare-option -docstring %{Whether to automatically update the buffer on changes} bool parinfer_enabled yes
+define-command -docstring "parinfer-enable-window [<mode>]: enable Parinfer for current window.
+Modes:
+    -indent  Preserve indentation and fix parentheses (default).
+    -paren   Preserve parentheses and fix indentation.
+    -smart   Try to be smart about what to fix." \
+parinfer-enable-window -params ..1 %{
+    require-module parinfer
+    try %{
+        parinfer -paren
+        # if parinfer -paren fails, we don't set parinfer_enable to true
+        # thus making sure that hooks do not break our code
+        set-option window parinfer_enabled true
+    } catch %{
+        # set up recovery hooks that will re-enable parinfer when parens are balanced
+        hook -group parinfer-try-paren window NormalIdle .* parinfer-try-paren
+        hook -group parinfer-try-paren window InsertIdle .* parinfer-try-paren
+        echo -markup "{Error}parinfer error: %val{error}"
+        echo -debug "parinfer error: %val{error}"
+    }
+    evaluate-commands %sh{
+        mode="${1:-indent}"
+        printf "%s\n" "remove-hooks window parinfer
+                       hook -group parinfer window NormalKey .* %{ parinfer -if-enabled $mode }
+                       hook -group parinfer window InsertChar (?!\n).* %{ parinfer -if-enabled $mode }
+                       hook -group parinfer window InsertDelete .* %{ parinfer -if-enabled $mode }"
+    }
+}
+
+define-command -docstring "parinfer-disable-window: disable Parinfer for current window." \
+parinfer-disable-window %{
+    remove-hooks window parinfer
+    remove-hooks window parinfer-try-paren
+    set-option window parinfer_enabled false
+}
 
 provide-module parinfer %{
 
+declare-option -docstring "Whether to automatically update the buffer on changes" \
+bool parinfer_enabled false
+
+declare-option -docstring "Currently Parinfer active mode" \
+str parinfer_current_mode
+
 declare-option -hidden str parinfer_previous_text
 declare-option -hidden str parinfer_previous_cursor_char_column
 declare-option -hidden str parinfer_previous_cursor_line
 declare-option -hidden str parinfer_previous_timestamp
 declare-option -hidden int parinfer_cursor_char_column
 declare-option -hidden int parinfer_cursor_line
 
-define-command -params .. \
-    -docstring %{parinfer [<switches>]: reformat buffer with parinfer-rust
-Modes:
-    -indent  Preserve indentation and fix parentheses (default).
-    -paren   Preserve parentheses and fix indentation.
-    -smart   Try to be smart about what to fix.} \
-    parinfer %{
-    eval -draft -save-regs '/"|^@' -no-hooks %{
+define-command -hidden -docstring "parinfer [<switches>]: reformat buffer with parinfer-rust.
```

Comment:
`parinfer` should not be hidden, it can be used one-off to reformat a buffer.  (If it were to be hidden, we should remove the docstring.)

Reply:
fixed. Also what do you think on providing a function `parinfer-set-indent`? It will set correct indentation for current line based on AST, and thus could be mapped to the <kbd>Tab</kbd>